### ISSUE TITLE
Creation of linked variations doesn't work for a big amount of attribute values

### ISSUE THREAD

**Issue by Braintelligence** [woocommerce/woocommerce#22639]
**Describe the bug**
Hi everyone. I have a problem with a lot of attributes while creating linked variations.

Normally, when you start creating variants from attributes and you see a progression, you can already see part-results in another tab before it's even done.

I have a product with a really big number of attribute-choices and when trying to create variants from them I see the loading symbol but nothing happens at all.

I'm not sure what could be limiting hits, I also don't know where I should look for an error log here.

Anyone have an advice for what I should try to look for in fixing this? I can throw a lot more memory and execution time at this, if needed. Perhaps I'm missing the correct setting.

I can't find the bottleneck of this function: https://github.com/woocommerce/woocommerce/blob/40694118fc1a44aea7b5a88edeaf37fc291c0ab9/includes/class-wc-ajax.php#L690
Raising max_input_vars and max_input_nesting_level didn't help.

**To Reproduce**
1. Create a product with at least 6 attributes and at least 20 possible values per attribute.
2. Try to create linked variations.

**Screenshots**
Just so you have a rough idea of what we're talking about here are some of those attributes:
![grafik](https://user-images.githubusercontent.com/6761196/52166677-a011ec80-2710-11e9-9925-ed1ec4b5b3d3.png)

**Expected behavior**
I'd expect creation of at least one linked variation after a while. Instead you see the loading icon endlessly and nothing happens. I tried to lessen the amount of attribute values and hit a point where everything worked. The difference between relatively swift creation of linked variations and **nothing** was one attribute value more or less at that point.

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active. (The report contains the plugins I want to use, just in case)
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**WordPress Environment**
<details>
### WordPress Environment ###

Home URL: https://adawoo.ravok.de
Site URL: https://adawoo.ravok.de
WC Version: 3.5.4
Log Directory Writable: ✔
WP Version: 5.0.3
WP Multisite: ✔
WP Memory Limit: 256 MB
WP Debug Mode: –
WP Cron: ✔
Language: de_DE_formal
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.25 (Debian)
PHP Version: 7.2.13
PHP Post Max Size: 256 MB
PHP Time Limit: 10000
PHP Max Input Vars: 1000
cURL Version: 7.52.1
OpenSSL/1.0.2q

SUHOSIN Installed: –
MySQL Version: 8.0.13
Max Upload Size: 63 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ❌ Ihr Server verfügt nicht über eine lauffähige Version der SoapClient Klasse - einige Zahlungsschnittstellen
welche SOAP nutzen
funktionieren unter Umständen nicht korrekt.

DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.5.4
WC Database Prefix: wp_
MaxMind GeoIP Database: ❌ Die MaxMind GeoIP Datenbank existiert nicht - Die Geolokalisierung (Ortung) wird nicht funktionieren. Sie können sie von https://dev.maxmind.com/geoip/geoip2/geolite2/ manuell herunterladen und in den Pfad:  installieren. Scrollen Sie dazu nach unten bis zu „Downloads“ und laden Sie die Datei „MaxMind DB binär
gzipped“ vom Typ „GeoLite2 Country“ herunter. Denken Sie bitte daran
die Datei GeoLite2-Country_xxxxxxxx.tar.gz zu entpacken und nur die entpackte Datei GeoLite2-Country.mmdb hochzuladen.

Datenbank-Gesamtgröße: 188.67MB
Datenbank-Datengröße: 183.18MB
Datenbank-Indexgröße: 5.49MB
wp_woocommerce_sessions: Daten: 0.02MB + Index: 0.02MB
wp_woocommerce_api_keys: Daten: 0.02MB + Index: 0.03MB
wp_woocommerce_attribute_taxonomies: Daten: 0.02MB + Index: 0.02MB
wp_woocommerce_downloadable_product_permissions: Daten: 0.02MB + Index: 0.05MB
wp_woocommerce_order_items: Daten: 0.02MB + Index: 0.02MB
wp_woocommerce_order_itemmeta: Daten: 0.02MB + Index: 0.03MB
wp_woocommerce_tax_rates: Daten: 0.02MB + Index: 0.06MB
wp_woocommerce_tax_rate_locations: Daten: 0.02MB + Index: 0.03MB
wp_woocommerce_shipping_zones: Daten: 0.02MB + Index: 0.00MB
wp_woocommerce_shipping_zone_locations: Daten: 0.02MB + Index: 0.03MB
wp_woocommerce_shipping_zone_methods: Daten: 0.02MB + Index: 0.00MB
wp_woocommerce_payment_tokens: Daten: 0.02MB + Index: 0.02MB
wp_woocommerce_payment_tokenmeta: Daten: 0.02MB + Index: 0.03MB
wp_woocommerce_log: Daten: 0.02MB + Index: 0.02MB
wp_aiowps_events: Daten: 0.02MB + Index: 0.00MB
wp_aiowps_failed_logins: Daten: 0.02MB + Index: 0.00MB
wp_aiowps_global_meta: Daten: 0.02MB + Index: 0.00MB
wp_aiowps_login_activity: Daten: 0.02MB + Index: 0.00MB
wp_aiowps_login_lockdown: Daten: 0.02MB + Index: 0.00MB
wp_aiowps_permanent_block: Daten: 0.02MB + Index: 0.00MB
wp_blog_versions: Daten: 0.02MB + Index: 0.02MB
wp_blogs: Daten: 0.02MB + Index: 0.03MB
wp_commentmeta: Daten: 0.02MB + Index: 0.03MB
wp_comments: Daten: 0.02MB + Index: 0.09MB
wp_links: Daten: 0.02MB + Index: 0.02MB
wp_options: Daten: 176.41MB + Index: 0.05MB
wp_postmeta: Daten: 3.52MB + Index: 4.00MB
wp_posts: Daten: 0.34MB + Index: 0.31MB
wp_registration_log: Daten: 0.02MB + Index: 0.02MB
wp_signups: Daten: 0.02MB + Index: 0.06MB
wp_site: Daten: 0.02MB + Index: 0.02MB
wp_sitemeta: Daten: 1.48MB + Index: 0.03MB
wp_social_users: Daten: 0.02MB + Index: 0.02MB
wp_term_relationships: Daten: 0.13MB + Index: 0.05MB
wp_term_taxonomy: Daten: 0.08MB + Index: 0.06MB
wp_termmeta: Daten: 0.06MB + Index: 0.06MB
wp_terms: Daten: 0.06MB + Index: 0.06MB
wp_usermeta: Daten: 0.20MB + Index: 0.03MB
wp_users: Daten: 0.02MB + Index: 0.05MB
wp_wc_download_log: Daten: 0.02MB + Index: 0.03MB
wp_wc_webhooks: Daten: 0.02MB + Index: 0.02MB
wp_yith_wcwl: Daten: 0.02MB + Index: 0.02MB
wp_yith_wcwl_lists: Daten: 0.02MB + Index: 0.03MB
wp_yoast_seo_links: Daten: 0.02MB + Index: 0.02MB
wp_yoast_seo_meta: Daten: 0.20MB + Index: 0.00MB

### Post Type Counts ###

attachment: 22
customize_changeset: 12
featured_item: 1
nav_menu_item: 2
page: 6
polylang_mo: 1
post: 3
product: 12
product_variation: 1329
wpcf7_contact_form: 1

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (22) ###

Contact Form 7: von Takayuki Miyoshi – 5.1.1
Jetpack von WordPress.com: von Automattic – 6.9
Klarna Payments for WooCommerce: von krokedil
klarna
automattic – 1.6.3

Nextend Social Login: von Nextendweb – 3.0.17
WooCommerce Blocks: von Automattic – 1.3.1
WooCommerce PayPal Checkout Gateway: von WooCommerce – 1.6.8
WooCommerce Stripe-Gateway: von WooCommerce – 4.1.14
WooCommerce Services: von Automattic – 1.18.0
WooCommerce: von Automattic – 3.5.4
WooSidebars: von WooCommerce – 1.4.5
YITH WooCommerce Wishlist: von YITH – 2.2.7
All In One WP Security: von Tips and Tricks HQ
Peter Petreski
Ruhul
Ivy – 4.3.8.3

Contact Form 7: von Takayuki Miyoshi – 5.1.1
YITH WooCommerce Wishlist: von YITH – 2.2.7
WooSidebars: von WooCommerce – 1.4.5
WooCommerce Blocks: von Automattic – 1.3.1
WooCommerce: von Automattic – 3.5.4
Polylang: von Frédéric Demarle – 2.5.1
WP File Manager: von mndpsingh287 – 4.1
W3 Total Cache: von Frederick Townes – 0.9.7.2
Yoast SEO: von Team Yoast – 9.5
WooCommerce UPC, EAN, and ISBN: von Scott Bolinger – 0.4.0 – Nicht getestest mit der aktiven WooCommerce-Version

### Settings ###

API Enabled: –
Force SSL: –
Currency: EUR (€)
Currency Position: left
Thousand Separator: .
Decimal Separator: ,
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

### WC Pages ###

Shop-Basis: #8 - /shop/
Warenkorb: #9 - /warenkorb/
Kasse: #10 - /kasse/
Mein Konto: #11 - /mein-konto/
Allgemeine Geschäftsbedingungen: ❌ Seite nicht festgelegt

### Theme ###

Name: Flatsome Child
Version: 3.0
Author URL: 
Child Theme: ✔
Parent Theme Name: Flatsome
Parent Theme Version: 3.7.2
Parent Theme Author URL: http://www.uxthemes.com/
WooCommerce Support: ✔

### Templates ###

Overrides: flatsome/woocommerce/archive-product.php
flatsome/woocommerce/cart/cart-empty.php
flatsome/woocommerce/cart/cart-shipping.php
flatsome/woocommerce/cart/cart.php
flatsome/woocommerce/checkout/form-checkout.php
flatsome/woocommerce/checkout/form-coupon.php
flatsome/woocommerce/checkout/thankyou.php
flatsome/woocommerce/content-product.php
flatsome/woocommerce/content-product_cat.php
flatsome/woocommerce/content-single-product.php
flatsome/woocommerce/content-widget-product.php
flatsome/woocommerce/global/breadcrumb.php
flatsome/woocommerce/global/quantity-input.php
flatsome/woocommerce/global/wrapper-end.php
flatsome/woocommerce/global/wrapper-start.php
flatsome/woocommerce/loop/loop-end.php
flatsome/woocommerce/loop/loop-start.php
flatsome/woocommerce/loop/pagination.php
flatsome/woocommerce/loop/result-count.php
flatsome/woocommerce/loop/sale-flash.php
flatsome/woocommerce/myaccount/form-login.php
flatsome/woocommerce/myaccount/navigation.php
flatsome/woocommerce/notices/error.php
flatsome/woocommerce/notices/notice.php
flatsome/woocommerce/notices/success.php
flatsome/woocommerce/product-searchform.php
flatsome/woocommerce/single-product/photoswipe.php
flatsome/woocommerce/single-product/price.php
flatsome/woocommerce/single-product/product-image.php
flatsome/woocommerce/single-product/product-thumbnails.php
flatsome/woocommerce/single-product/rating.php
flatsome/woocommerce/single-product/related.php
flatsome/woocommerce/single-product/review.php
flatsome/woocommerce/single-product/sale-flash.php
flatsome/woocommerce/single-product/share.php
flatsome/woocommerce/single-product/short-description.php
flatsome/woocommerce/single-product/tabs/tabs.php
flatsome/woocommerce/single-product/title.php
flatsome/woocommerce/single-product/up-sells.php
flatsome/woocommerce/single-product-reviews.php
flatsome/woocommerce/single-product.php
</details>

**Comment by rrennick** [comment]
The variation creation runs in the background with a server configured time limit. The creation process is terminated by the server at the end of the time limit. 

When that happens you can refresh the page & use the create variations from all again. It will pick up where it left off. With that number of variations it will take a few times to create all the variations.

Since this is a server resource issue I'm going to close the issue.

**Comment by Braintelligence** [comment]
@rrennick Wait, you assessed the situation wrongly.

I'll repeat: If I remove one single attribute value everything works fine. Creating, let's say 50, variations is done within a few seconds. If I add back a single attribute value to any attribute nothing happens for hours.

This is **not** a time-limit issue, rather the creation of linked variations is never initialized.

Please reopen this issue @rrennick

**Comment by rrennick** [comment]
@Braintelligence Can you provide details on steps to reproduce?

**Comment by Braintelligence** [comment]
Well as described in the initial post you need to create a product with absurdly many attribute values and at least 6 attributes (perhaps less, it seems that many attribute values seem to be the main issue).

This is a real use-case, though. The product in question is a contact lens with absurdly many possible combinations of attributes of which each has its own EAN.

The screenshot I provided in the initial post should give an idea of the attribute dimensions. There should come a point where creating linked variations will just do nothing but spawn a loading symbol. No visible error to be seen and I don't know where to find appropriate logs.

You should safely hit that spot if you use 6 attributes with 20 values each, I guess.

**Comment by rrennick** [comment]
In reviewing the steps to reproduce, the minimum number of attributes needed to reproduce this is 20 terms in each of 6 attributes. That is 20^6 = 64000000 variations. @Braintelligence is that what you are trying to achieve?

**Comment by Braintelligence** [comment]
@rrennick No that is not the amount of variations needed. It should be somewhat less than 10000. 
The phenomenon should just safely pop up with that amount of attribute values.
If it helps I can tell you the exact amount of attributes and values needed for me to be on the edge between working and not working.

**Comment by Braintelligence** [comment]
So I have the following variant-relevant attributes and the corresponding amount of unique values for my real use-case product (toric contact lens):
```
BC - 17
PWR - 104
CYL - 20
AXIS - 181
DIA - 26
ADD - 16
```
All attribute values are degrees or other numeric kind of values.

This doesn't work at all when creating linked variations.
Reducing the number of values made it work magically at some point, which first was:
```
BC - 1
PWR - 83
CYL - 2
AXIS - 161
DIA - 6
ADD - 1
```
by fiddling around a bit I hit this point where it didn't work anymore:
```
BC - 2
PWR - 81
CYL - 2
AXIS - 161
DIA - 6
ADD - 1
```
changing PWR to 80 makes it work in a few seconds. Setting it back to 81 makes it loop endlessly again.

Please help me making this work =(...

**Comment by rrennick** [comment]
> Could we possibly run into a `PHP_INT_MAX` problem here?

My first guess would be that it's exhausting the PHP memory limit. To create a temporary set of all combinations:

Works: 1 x 83 x 2 x 161 x 6 x 1 = 160356
Not:  2 x 83 x 2 x 161 x 6 x 1 = 320712

Your real use-case is 2662466560. WC allows you to use attributes without creating distinct variations so you could look at an alternative way of capturing the distinct data you need.

Closing as this is a resource issue.

**Comment by Braintelligence** [comment]
@rrennick I will try raising the PHP memory limit to a higher number. I have a lot of RAM to spare. Though iirc raising from 128M to 256M had zero effect.

You also again didn't capture the whole picture. My last example that still works is 

2 x 80 x 2 x 161 x 6 x 1 = 309120

It doesn't work with

2 x 81 x 2 x 161 x 6 x 1 = 312984

Can you elaborate on how you can explain the former to create 10 variants in a few seconds and the other one to do nothing by PHP memory limit? Why am I also not seeing an error saying PHP memory was exceeded after half an hour?

I'm not convinced that this is a ressource issue.

**Comment by rrennick** [comment]
> Though iirc raising from 128M to 256M had zero effect.

The 128MB memory limit is on the front end. The dashboard & WC ajax requests default to 256MB if the server resource limits allow it.

**Comment by Braintelligence** [comment]
Yeah so the following happened:
1. cranked up PHP memory limit to 116GB
2. when creating ~320000 variants I get a spike to 2GB RAM for the full system used and 12 variants are created almost instantly
3. when creating ~2662466560 variants I get a slowly building spike to >111GB RAM !!! used, it drops down to 50GB RAM then and CPU usage on 4 cores drops from 20~40% to 2%. Then RAM usage drops to 25GB still nothing to be seen on frontend. 25GB RAM keep being used.
![grafik](https://user-images.githubusercontent.com/6761196/52345707-29316800-2a1e-11e9-997b-c912bb13f63d.png)

When you think about it, if I need 256M memory for ~320k variants I will need about ~2129973M memory for ~2662466560 variants. Which is more than 2 TERABYTE RAM. This can't be ok.

No way in hell an efficient algorithm would need that much RAM for something as simple as 6 nested for-loops with even ~200 calls each on trivial datachunks.
The fact that I can't make this work via Woocommerce even with 116GB RAM but can make it work with a tiny fraction of this RAM via a simple Python-Script to generate a huge CSV should convert this issue from a ressource problem on my side to an algorithm efficiency issue on this repo's side.

This is as a warning for anyone trying to attempt to do the same. You need to throw a metric ton of ressources at Woocommerce if you want to convey some real use-cases into product variants here. If you need to do something like I did, do it via a CSV-import instead.
